x/mongo/driver/operation.go

```diff
@@ -1078,6 +1094,28 @@ func (op Operation) addServerAPI(dst []byte) []byte {
 	return dst
 }
 
+// addMaxTimeMS adds a 'maxTimeMS' field to the wire message in dst based on the current context's deadline
+// and the 90th percentile RTT if op.Timeout is set.
+func (op Operation) addMaxTimeMS(ctx context.Context, dst []byte, desc description.SelectedServer) []byte {
+	if op.Timeout != nil {
+		// Retrieve current deadline from context, convert it to timeout and append
+		// maxTimeMS as (timeout - 90th percentile RTT).
+		if deadline, deadlineSet := ctx.Deadline(); deadlineSet {
+			remainingTimeout := time.Until(deadline)
+			// If 90th percentile RTT is less than remainingTimeout, error instead
+			// of attempting to send message to server.
+			// TODO(CSOT): Uncomment this once RTT90 is actually being calculated.
+			// if desc.RTT90 < remainingTimeout {
+			//	return dst, internal.WrapErrorf(ErrDeadlineWouldBeExceeded,
+			//		"Remaining timeout %v applied from Timeout is less than 90th percentile RTT", remainingTimeout)
+			// }
+			maxTimeMS := int64(remainingTimeout/time.Millisecond) - int64(desc.RTT90/time.Millisecond)
+			dst = bsoncore.AppendInt64Element(dst, "maxTimeMS", maxTimeMS)
```

Comment:
The combination of client-side and server-side timeouts may create a situation where a the same operation call returns different timeout messages depending on driver- and server-side conditions.

For example, consider we're running this logic near the timeout deadline. If the network round-trip is actually slower than predicted, the Go process may trigger its timeout condition first, resulting in a client-side timeout error. If the network round-trip is actually faster than predicted, the server may trigger its timeout condition first, resulting in a server-side timeout error. If the Go process is overloaded and our operation goroutine doesn't run on time, we may short-circuit resulting in a different client-side timeout error. That variability in errors may cause confusion from users.

Is the main goal for the query to always reach `maxTimeMS` first and let the server send an error response back before the client-side timeout, allowing connection reuse? Or is the main goal to limit server-side resource use when the result is no longer needed? If the goal is the former, we will need to instruct users how to handle the mix of client-side and server-side timeout errors. If the goal is the latter, we should consider setting a longer `maxTimeMS` to minimize the probability that we get mixed client-side and server-side errors.

Reply:
`maxTimeMS` is difficult to compare to the timeout in the driver because they measure different things. `maxTimeMS` measures time spent processing where the driver timeout measures wall clock duration. For example (assuming zero network latency for simplicity), it's possible to run a Find with a client-side timeout of 1s and `maxTimeMS` of 500ms and reach the client-side timeout first because the server didn't spend 500ms actively processing the cursor in the 1s wall clock time. Basically the wall clock time it takes to reach `maxTimeMS` will never be less than the specified time, but could be more under some circumstances. As a result, even if `maxTimeMS` < `remainingTimeout`, it's possible to reach `remainingTimeout` first.

Check out the note in the manual page for [cursor.maxTimeMS()](https://docs.mongodb.com/manual/reference/method/cursor.maxTimeMS/#mongodb-method-cursor.maxTimeMS):
> A cursor's idle time does not contribute towards its processing time.
https://docs.mongodb.com/manual/reference/method/cursor.maxTimeMS/#mongodb-method-cursor.maxTimeMS
Also check out this [StackOverflow question](https://stackoverflow.com/questions/24964708/mongo-maxtimems-doenst-kill-query-immediately) about `maxTimeMS` for an example of how that difference can manifest.

Reply:
Ah great point; I didn't realize `maxTimeMS` did not consider the idle time of the cursor in its calculation of processing time. It sounds like users will certainly see a mix of client-side and server-side timeout errors, then. I _could_ transform server-side errors from exceeded `maxTimeMS` into `context.DeadlineExceeded` errors, but that error manipulation feels odd. I've filed GODRIVER-2339 to think about the new error API for CSOT separately.

Reply:
There are two open questions here:
1. How do we make handling different errors that all represent a timeout ergonomic?
2. Does the intended improvement of setting `maxTimeMS` actually happen in practice?

GODRIVER-2339 covers question 1. However, for question 2 I think we need to set up some test scenarios and measure the impact of using `maxTimeMS` (and maybe also skipping operations). Based on the rationale listed in the [`maxTimeMS` accounts for server RTT](https://github.com/mongodb/specifications/blob/master/source/client-side-operations-timeout/client-side-operations-timeout.rst#maxtimems-accounts-for-server-rtt) section, it seems like the primary reason for using `maxTimeMS` is to limit connection churn. We should set up a few scenarios to test if that expected reduction in connection churn actually happens.

Suggested scenarios:
* Primarily network latency - Connect to a MongoDB deployment with an average RTT of >500ms. Run `FindOne` operations that match a document on an indexed field with a timeout that is double the average RTT. Run operations in a loop in 100 goroutines.
* Primarily server latency - Connect to a MongoDB deployment with an average RTT of <10ms. Run `Aggregate` operations on a collection with lots of items that takes 10+ seconds to complete. Run operations in a loop in 10 goroutines.

Then measure the number of connections closed and created when using `maxTimeMS` and not using `maxTimeMS`. Expect that there are fewer connections closed and created when using `maxTimeMS`.